fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/PrestoSqlBackedClient.java

```diff
@@ -0,0 +1,259 @@
+                    }
+                    return null;
+                }
+        );
+    }
+
+    /**
+     * Builds sql for a druid query, convert it to presto dialect,
+     * execute it against the database, process the results
+     * and return a jsonNode in the format of a druid response.
+     *
+     * @param druidQuery  The druid query to build and process.
+     *
+     * @return a druid-like response to the query.
+     */
+    private JsonNode executeAndProcessQuery(DruidAggregationQuery<?> druidQuery) {
+        if (!druidQueryToPrestoConverter.isValidQuery(druidQuery)) {
+            throw new UnsupportedOperationException("Unable to process " + druidQuery);
+        }
+
+        ApiToFieldMapper aliasMaker =
+                new ApiToFieldMapper(druidQuery.getDataSource().getPhysicalTable().getSchema());
+        String sqlQuery = druidQueryToPrestoConverter.buildSqlQuery(druidQuery, aliasMaker);
+
+        LOG.info("Input raw sql query: {}", sqlQuery);
+        sqlQuery = sqlQueryToPrestoQuery(sqlQuery);
+        LOG.info("Processed to presto query: {}", sqlQuery);
+
+        PrestoResultSetProcessor resultSetProcessor = new PrestoResultSetProcessor(
+                druidQuery,
+                aliasMaker,
+                jsonWriter,
+                druidQueryToPrestoConverter.getTimeConverter()
+        );
+
+        try (Connection connection = calciteHelper.getConnection()) {
+            Statement statement = connection.createStatement();
+            ResultSet resultSet = statement.executeQuery(sqlQuery);
+            resultSetProcessor.process(resultSet);
+            JsonNode jsonNode = resultSetProcessor.buildDruidResponse();
+            LOG.info("Created response: {}", jsonNode);
+            return jsonNode;
+        } catch (SQLException e) {
+            LOG.warn("Failed while processing {}", druidQuery);
+            throw new RuntimeException("Couldn't generate sql", e);
+        }
+    }
+
+    /**
+     * Converts sql dialect into presto compatible dialect. The input sqlQuery must follow the format of
+     * the output of DruidQueryToPrestoConverter.buildSqlQuery()
+     * Steps are:
+     * <ul>
+     *     <li>1. Replace Time functions with String Functions to be used for HIVE tables</li>
+     *     <li>2. Convert datestamp from Date format to String format</li>
+     *     <Li>3. Specify Catalog in FROM expression</Li>
+     *     <li>4. Calcite wrap column name in single quotes which is not allowed in Presto
+     *            as Presto is not ANSI compliant.
+     *            Replace single quotation marks with double quotation marks.</li>
+     *     <li>5. Convert FETCH NEXT n ROWS expression to prestodb-supported LIMIT n expression</li>
+     * </ul>
+     *
+     * @param sqlQuery The sql query converted from druid query.
+     *
+     * @return a presto compatible sql dialect.
+     * */
+    private static String sqlQueryToPrestoQuery(String sqlQuery) {
```

Comment:
Can we break this up into the logical steps and then make this 
```suggestion
    protected String sqlQueryToPrestoQuery(String sqlQuery) {
```
so that this can call each of the transforms and be overridable.

Then you can leave the date/time fixes in your repository as those are a result of storing the timestamp as a string (see [timestamp restriction](https://github.com/yahoo/fili/tree/master/fili-sql#notable-restrictions)) so it shouldn't apply to everyone. It would be nice to provide better support for having your own year/month/etc grain parsing but this solution should work for now

Reply:
Discuss

Reply:
Fixed.